### ISSUE TITLE
Question - Difference using `t` helper and `intl.t` from the Service

### ISSUE THREAD

**Issue by MrChocolatine** [ember-intl/ember-intl#1441]
- [x] I have searched the [issues](https://github.com/ember-intl/ember-intl/issues) of this repo and believe that this is not a duplicate

### Environment

- Ember Version: 3.16.10
- Ember CLI Version: 3.16.2
- Ember Intl Version: 4.4.1

### Question

Hello there, I have a small question.

I am wondering, is there a difference between using `t` macro helper  and `intl.t` from the Service?

https://ember-intl.github.io/ember-intl/docs/guide/translating-text#t

```js
import { t } from 'ember-intl';

export default class ... {
  title = t('my.key');
});
```

https://ember-intl.github.io/ember-intl/docs/guide/service-api#t-translationkey-string-optionaloptions-object-string-safestring

```js
export default class ... {
  @service intl;

  constructor() {
    title = this.intl.t('my.key');
  }
});
```

For some cases I need `@computed('intl.locale')` so obviously the Service will be injected and I will use `this.intl.t()`.

Thanks for your explanation.

**Comment by buschtoens** [comment]
Hey! Yes, there is a difference.

First of all, I think your first code example, would not work:

> ```js
> import { t } from 'ember-intl';
> 
> export default class ... {
>   title = t('my.key');
> };
> ```

The issue is that computed property macros cannot be used via assignment on native classes, like you've shown above. Only the legacy Ember Object Model supports this. It'd look like this:

```ts
import { t } from 'ember-intl';

export default EmberObject.extend({
  title: t('my.key')
});
```

With native classes you have to use decorators. To be honest, I don't know / have not checked yet, whether the standard `ember-intl` `t` macro can be used as a decorator - there is a chance that this could work by now, since there's been some Ember-internal changes:

```ts
import { t } from 'ember-intl';

export default class {
  @t('my.key') title;
}
```

If it does not, [`@ember-intl/decorators`](https://github.com/ember-intl/decorators) would definitely work, as we wrote it specifically for that purpose.

```ts
import Component from '@ember/component';
import { t } from '@ember-intl/decorators';

class ExampleComponent extends Component {
  name = 'Tom';

  @t('messages.welcome', { firstName: 'name' }) message; // => 'Welcome, Tom!'
}
```

But now onto your actual question: What is the difference between using the `t` computed property macro and assigning the return value from the `t` service method?

When you assign the value (`title = this.intl.t('my.key')`), the assignment happens _once_, when the class instance is initialized. If you then change the locale, the value will not update. However, if you use the macro, the value _will_ update, when the locale is changed.

I hope this answers your question!

**Comment by buschtoens** [comment]
Prompted by this and the recent modernizations flowing into `ember-intl`, I think we should do two things though:

- Make sure the `t` macro can be used as a decorator or rather basically merge `@ember-intl/decorators`  into `ember-intl`
- Ensure that using `this.intl.t('...')` in an auto-tracking context correctly entangles the `locale` as a dependency, so that changing the locale invalidates the getter

Thanks for raising this!

**Comment by MrChocolatine** [comment]
I did not know about [ember-intl/decorators](https://github.com/ember-intl/decorators), that's really cool.

> When you assign the value (title = this.intl.t('my.key')), the assignment happens once, when the class instance is initialized. If you then change the locale, the value will not update. However, if you use the macro, the value will update, when the locale is changed.

Good to know ! Like I said previously, sometimes I need `@computed('intl.locale')` so I inject the Service and I use `this.intl.t()`, here is a simplified example:

```js
@service intl;
...
@computed('intl.locale')
get tableColumns() {
  return [
    {
      label: this.intl.t('my.key.aa'),
      ...
    },
    {
      label: this.intl.t('my.key.bb'),
      ...
    }
  ]
}
```

So, according to what to you said, could it be refactored like this?

```js
import { t } from 'ember-intl';
...
get tableColumns() {
  return [
    {
      label: t('my.key.aa'),
      ...
    },
    {
      label: t('my.key.bb'),
      ...
    }
  ]
}
```

Your answers were really helpful, thanks a lot.

**Comment by buschtoens** [comment]
> So, according to what to you said, could it be refactored like this?
> 
> ```js
> import { t } from 'ember-intl';
> ...
> get tableColumns() {
>   return [
>     {
>       label: t('my.key.aa'),
>       ...
>     },
>     {
>       label: t('my.key.bb'),
>       ...
>     }
>   ]
> }
> ```

That wouldn't work, because you're returning the computed property macro, not the value it would resolve to.

You _could_ use the [`@intl` decorator](https://github.com/ember-intl/decorators#intl):

```ts
import Component from '@glimmer/component';
import { intl } from '@ember-intl/decorators';

class ExampleComponent extends Component {
  @intl
  tableColumns = intl => [
    { label: intl.t('my.key.aa') },
    { label: intl.t('my.key.bb') }
  ];
}
```

But tbh, I'd probably just stick with injecting the `intl` service and using it directly, like you already do.

What we should check however, at least for modern Ember Octane apps, is:

>     * Ensure that using `this.intl.t('...')` in an auto-tracking context correctly entangles the `locale` as a dependency, so that changing the locale invalidates the getter

This way you could remove the `@computed('intl.locale')`.

**Comment by MrChocolatine** [comment]
> That wouldn't work, because you're returning the computed property macro, not the value it would resolve to.

Ah yes, right. It seems I easily forget it is a **computed property macro** so indeed it won't return the value.

Again many thanks, great explanations. It also helped other JS dev of my team.

**Comment by jasonmit** [comment]
> Ensure that using this.intl.t('...') in an auto-tracking context correctly entangles the locale as a dependency, so that changing the locale invalidates the getter

I believe there was an attempt to resolve this: https://github.com/ember-intl/ember-intl/pull/1340

Is that not working - landed in v5.2.0
